Unsize coercion fails if end of method is unreachable

The following async trait fails to compile. Notice that the same method body is fine in an ordinary async fn outside of an async trait.

This worked prior to #143.

```rust
use async_trait::async_trait;

struct Thing;
trait Ret {}
impl Ret for Thing {}

async fn ok() -> &'static dyn Ret {
    return &Thing;
}

#[async_trait]
trait Trait {
    async fn fail() -> &'static dyn Ret {
        return &Thing;
    }
}
```

```console
error[E0308]: mismatched types
  --> src/main.rs:13:41
   |
13 |       async fn fail() -> &'static dyn Ret {
   |  _________________________________________^
14 | |         return &Thing;
15 | |     }
   | |_____^ expected struct `Thing`, found trait object `dyn Ret`
   |
   = note:   expected type `&Thing`
           found reference `&'static (dyn Ret + 'static)`
note: return type inferred to be `&Thing` here
  --> src/main.rs:14:16
   |
14 |         return &Thing;
   |                ^^^^^^

error[E0271]: type mismatch resolving `<impl Future as Future>::Output == &'static (dyn Ret + 'static)`
  --> src/main.rs:13:41
   |
13 |       async fn fail() -> &'static dyn Ret {
   |  _________________________________________^
14 | |         return &Thing;
15 | |     }
   | |_____^ expected trait object `dyn Ret`, found struct `Thing`
   |
   = note: expected reference `&'static (dyn Ret + 'static)`
              found reference `&Thing`
   = note: required for the cast to the object type `dyn Future<Output = &'static (dyn Ret + 'static)> + Send`
```

--- a/src/expand.rs
+++ b/src/expand.rs
@@ -345,27 +345,21 @@ fn transform_block(sig: &mut Signature, block: &mut Block) {
     }
 
     let stmts = &block.stmts;
-    let ret = match &sig.output {
-        ReturnType::Default => quote!(()),
-        ReturnType::Type(_, ret) => quote!(#ret),
-    };
     let let_ret = match &sig.output {
         ReturnType::Default => quote_spanned! {block.brace_token.span=>
             let _: () = { #(#decls)* #(#stmts)* };
         },
         ReturnType::Type(_, ret) => quote_spanned! {block.brace_token.span=>
+            if let ::core::option::Option::Some(__ret) = ::core::option::Option::None::<#ret> {
+                return __ret;
+            }
             let __ret: #ret = { #(#decls)* #(#stmts)* };
             #[allow(unreachable_code)]
             __ret
         },
     };
     let box_pin = quote_spanned!(block.brace_token.span=>
-        Box::pin(async move {
-            if let ::core::option::Option::Some(__ret) = ::core::option::Option::None::<#ret> {
-                return __ret;
-            }
-            #let_ret
-        })
+        Box::pin(async move { #let_ret })
     );
     block.stmts = parse_quote!(#box_pin);
 }

--- a/src/expand.rs
+++ b/src/expand.rs
@@ -345,6 +345,10 @@ fn transform_block(sig: &mut Signature, block: &mut Block) {
     }
 
     let stmts = &block.stmts;
+    let ret = match &sig.output {
+        ReturnType::Default => quote!(()),
+        ReturnType::Type(_, ret) => quote!(#ret),
+    };
     let let_ret = match &sig.output {
         ReturnType::Default => quote_spanned! {block.brace_token.span=>
             let _: () = { #(#decls)* #(#stmts)* };
@@ -356,7 +360,12 @@ fn transform_block(sig: &mut Signature, block: &mut Block) {
         },
     };
     let box_pin = quote_spanned!(block.brace_token.span=>
-        Box::pin(async move { #let_ret })
+        Box::pin(async move {
+            if let ::core::option::Option::Some(__ret) = ::core::option::Option::None::<#ret> {
+                return __ret;
+            }
+            #let_ret
+        })
     );
     block.stmts = parse_quote!(#box_pin);
 }

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -1258,3 +1258,23 @@ pub mod issue147 {
         }
     }
 }
+
+// https://github.com/dtolnay/async-trait/issues/149
+pub mod issue149 {
+    use async_trait::async_trait;
+
+    pub struct Thing;
+    pub trait Ret {}
+    impl Ret for Thing {}
+
+    pub async fn ok() -> &'static dyn Ret {
+        return &Thing;
+    }
+
+    #[async_trait]
+    pub trait Trait {
+        async fn fail() -> &'static dyn Ret {
+            return &Thing;
+        }
+    }
+}

